fw/mmap_buffer.c

```diff
@@ -0,0 +1,305 @@
+/**
+ *		Tempesta FW
+ *
+ * Handling ring buffers is_ready to user space.
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "mmap_buffer.h"
+#include "lib/str.h"
+#include <linux/types.h>
+#include <linux/log2.h>
+#include <linux/cpumask.h>
+#include <linux/fs.h>
+#include <linux/device.h>
+#include <linux/io.h>
+#include <asm/page.h>
+
+/*
+ * We can't pass TfwMmapBufferHolder pointer to the file operations handlers.
+ * Let's store these pointers, and find them by filenames in the open handler.
+ */
+#define MAX_HOLDERS 4
+static TfwMmapBufferHolder *holders[MAX_HOLDERS];
+static int holders_cnt;
+
+static int dev_file_open(struct inode *ino, struct file *filp);
+static int dev_file_close(struct inode *ino, struct file *filp);
+static int dev_file_mmap(struct file *filp, struct vm_area_struct *vma);
+static void dev_file_vm_close(struct vm_area_struct *vma);
+
+static const struct file_operations dev_fops = {
+	.open  = dev_file_open,
+	.release  = dev_file_close,
+	.mmap  = dev_file_mmap,
+};
+
+static const struct vm_operations_struct dev_vm_ops = {
+	.close = dev_file_vm_close
+};
+
+void
+tfw_mmap_buffer_get_room(TfwMmapBufferHolder *holder,
+						 char **part1, unsigned int *size1,
+						 char **part2, unsigned int *size2)
+{
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+	u64 head, tail;
+
+	*size2 = 0;
+
+	if (!atomic_read(&buf->is_ready)) {
+		*size1 = 0;
+		return;
+	}
+
+	head = buf->head % buf->size;
+	tail = smp_load_acquire(&buf->tail) % buf->size;
+
+	*part1 = buf->data + head;
+
+	if (head < tail) {
+		*size1 = tail - head - 1;
+		return;
+	}
+
+	if (unlikely(head == 0)) {
+		*size1 = buf->size - 1;
+	} else {
+		*size1 = buf->size - head;
+		*part2 = buf->data;
+		*size2 = tail - 1;
+	}
+}
+
+void
+tfw_mmap_buffer_commit(TfwMmapBufferHolder *holder, unsigned int size)
+{
+	TfwMmapBuffer *buf = *this_cpu_ptr(holder->buf);
+
+	smp_store_release(&buf->head, buf->head + size);
+}
+
+static int
+dev_file_open(struct inode *ino, struct file *filp)
+{
+	TfwMmapBufferHolder *holder;
+	int i;
+
+	for (i = 0; i < holders_cnt; ++i) {
+		if (strcmp(holders[i]->dev_name, (char *)filp->f_path.dentry->d_iname) == 0) {
```

Comment:
too wide line: we that's OK to slightly exceed 80 characters per line, but not so much

Reply:
Fixed.